packages/react/src/suppress-safari-class-name-mismatch.js

```diff
@@ -0,0 +1,105 @@
+function classNameBelongsToCache(className, cacheKey) {
+  return className.startsWith(cacheKey + '-')
+}
+
+function getLabelFromClassName(className, cacheKey) {
+  const withoutCacheKey = className.replace(cacheKey + '-', '')
+  const parts = withoutCacheKey.split('-')
+
+  return parts.length === 2 ? parts[1] : undefined
+}
+
+export function isPossiblyBenignClassNameMismatch(
+  cache,
+  serverClassName,
+  browserClassName
+) {
+  const serverLabel = getLabelFromClassName(serverClassName, cache.key)
+  const browserLabel = getLabelFromClassName(browserClassName, cache.key)
```

Comment:
if we have access to the cache here - we could try to get both styles from the cache and rehash the browser styles with the label and compare the results. We should try to still passthrough the error message if the mismatch if legit

Reply:
Reading the code, it looks like serialized styles are only stored in the cache in the browser if the JSX element type is a component. I.e. styles are not stored in the cache for plain elements like `<div>`.

So this approach doesn't really seem worth it since the majority of the time, the styles will not be in the cache. Let me know if I am misunderstanding something.

Reply:
You are right. This prevents us from using the proposed strategy and I think it makes the whole thing not a viable solution because muting the possibly valid mismatch errors would be very bad. I wonder though if we couldn't just always store this in the development mode - where the memory concern shouldn't be as big of a problem. This is the part of the code that we are discussing:
https://github.com/emotion-js/emotion/blob/f3c51e8aca287170e43110bee2e4527eacfcada4/packages/utils/src/index.js#L29-L44

WDYT @mitchellhamilton ?

Reply:
That seems like a good idea. It seems like the memory usage is the only potential problem. I'll leave deciding that to you and Mitchell.

Here's another idea I had which could be a backup option.

-----------------------

On the server, Emotion adds a `style` tag with a `data-emotion` attribute, e.g.:

```html
<style data-emotion="css 18gru7x-Home">
  .css-18gru7x-Home {
    color: orchid;
  }
</style>
```

18gru7x is the hash including the label "Home".

We could change Emotion's server-side behavior so that it also computes the hash for the styles _without the label_. Say this results in the hash "abcdef". We would add this to the style tag like:

```html
<style data-emotion="css 18gru7x-Home abcdef">
  .css-18gru7x-Home {
    color: orchid;
  }
</style>
```

Then if I get a class name mismatch for `css-18gru7x-Home` and `css-abcdef`, I can find the appropriate style tag to determine if the mismatch is caused by failed label extraction in the browser.

Reply:
The `css-abcdef` rule would also have to be generated and put in the rendered `<style/>`. Those IDs are being treated by the client as already inserted ones - and the client doesn't ever try to re-insert what is already inserted. So without that, we'd risk that some component would get its styling because `css-abcdef` could be used as a class name somewhere but there would be no rules inserted for it.

But overall, this is a valid approach that could be taken - at the cost of generating more rules in dev (and that doesn't seem to be overly bad 🤔 ). 

I wonder though... there would still be cases, thanks to TCO, where the client and the server would both get a label - just a different one, right? Is there any mitigation that we could think of to address this? The rehashing proposal could potentially address this problem too - if only implemented robustly enough.

@@mitchellhamilton friendly 🏓

Reply:
It seems like we have two valid approaches here:

1. In the browser, store all serialized styles in the cache in dev only. Downside: more memory usage in browser.
2. On the server, hash everything with & without the label and insert style rules for both. Downside: more CPU usage during SSR and twice as many style elements in the HTML.

I'm feeling like option 1 is better because it is invisible to the developer, and Emotion's behavior will be more similar between dev & prod.

If Mitchell isn't available to provide feedback, I'll probably implement option 1 when I'm off for Christmas.

Reply:
I think a bit more memory usage in dev would be fine. Though, I've been thinking about this and a bigger concern I have is that caches couldn't be garbage collected which wouldn't be a concern for most people but could be quite problematic for people doing web component or iframe stuff. Maybe we could fix that by keeping the cache in a WeakRef for the usage in overriding console.error though? (I'm also kind of thinking, oh gosh, is this really worth it though?)

Reply:
> Maybe we could fix that by keeping the cache in a WeakRef for the usage in overriding console.error though? (I'm also kind of thinking, oh gosh, is this really worth it though?)

Probably not worth it since it's a dev-only path. I see your point but I find it hard to imagine that this will be a practical issue for anyone.